x/storage/keeper/cross_app_executor.go

```diff
@@ -0,0 +1,274 @@
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.pMsgServer.Deposit(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgTypeDisableRefund:
+		var gnfdMsg paymentmoduletypes.MsgDisableRefund
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.pMsgServer.DisableRefund(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgWithdraw:
+		var gnfdMsg paymentmoduletypes.MsgWithdraw
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.pMsgServer.Withdraw(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgMigrateBucket:
+		var gnfdMsg types.MsgMigrateBucket
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.MigrateBucket(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgCancelMigrateBucket:
+		var gnfdMsg types.MsgCancelMigrateBucket
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.CancelMigrateBucket(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgUpdateBucketInfo:
+		var gnfdMsg types.MsgUpdateBucketInfo
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.UpdateBucketInfo(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgToggleSPAsDelegatedAgent:
+		var gnfdMsg types.MsgToggleSPAsDelegatedAgent
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.ToggleSPAsDelegatedAgent(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgSetBucketFlowRateLimit:
+		var gnfdMsg types.MsgSetBucketFlowRateLimit
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.SetBucketFlowRateLimit(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgCopyObject:
+		var gnfdMsg types.MsgCopyObject
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
+			return fmt.Errorf("invalid msg sender")
+		}
+		_, err = app.sMsgServer.CopyObject(sdk.WrapSDKContext(ctx), &gnfdMsg)
+	case MsgUpdateObjectInfo:
+		var gnfdMsg types.MsgUpdateObjectInfo
+		err = gnfdMsg.Unmarshal(msg.Data)
+		if err != nil {
+			return err
+		}
+		if !gnfdMsg.GetSigners()[0].Equals(msgSender) {
```

Comment:
every handler has: `if !gnfdMsg.GetSigners()[0].Equals(msgSender) ` check, can we reuse the code?

Reply:
Done